enclave-runtime/src/sidechain_impl.rs

```diff
@@ -0,0 +1,268 @@
+//! Implement sidechain traits that can't be implemented outside.
+//!
+//! Todo: Once we have put the `top_pool` stuff in an entirely different crate we can
+//! move most parts here to the sidechain crate.
+
+use crate::{
+	exec_tops, prepare_and_send_xts_and_block,
+	rpc::author::{alloc::sync::Arc, AuthorApi, OnBlockCreated, SendState},
+	state::load,
+	Result as EnclaveResult,
+};
+use codec::Encode;
+use core::time::Duration;
+use itc_light_client::{BlockNumberOps, LightClientState, NumberFor, Validator};
+use itp_ocall_api::EnclaveOnChainOCallApi;
+use itp_settings::sidechain::SLOT_DURATION;
+use itp_sgx_crypto::{Aes, AesSeal};
+use itp_sgx_io::SealedIO;
+use itp_storage_verifier::GetStorageVerified;
+use its_sidechain::{
+	aura::{Aura, AuraVerifier, SlotClaimStrategy},
+	consensus_common::{BlockImport, Environment, Error as ConsensusError, Proposal, Proposer},
+	primitives::traits::{Block as SidechainBlockT, ShardIdentifierFor, SignedBlock},
+	slots::{PerShardSlotWorkerScheduler, SlotInfo},
+	state::SidechainDB,
+	validateer_fetch::ValidateerFetch,
+};
+use log::error;
+use primitive_types::H256;
+use sgx_externalities::SgxExternalities;
+use sp_core::Pair;
+use sp_runtime::{traits::Block, MultiSignature};
+use sp_std::prelude::Vec;
+use std::{marker::PhantomData, string::ToString};
+
+///! `SlotProposer` instance that has access to everything needed to propose a sidechain block
+pub struct SlotProposer<PB: Block, SB: SignedBlock, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub proposer_key: Pair,
+	pub parentchain_header: PB::Header,
+	pub shard: ShardIdentifierFor<SB>,
+	_phantom: PhantomData<PB>,
+}
+
+///! `ProposerFactory` instance containing all the data to create the `SlotProposer` for the
+/// next `Slot`
+pub struct ProposerFactory<PB: Block, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub pair: Pair,
+	_phantom: PhantomData<PB>,
+}
+
+impl<PB: Block, Pair, OcallApi, LightClient, Author>
+	ProposerFactory<PB, Pair, OcallApi, LightClient, Author>
+{
+	pub fn new(
+		ocall_api: Arc<OcallApi>,
+		light_client: Arc<LightClient>,
+		author: Arc<Author>,
+		pair: Pair,
+	) -> Self {
+		Self { ocall_api, light_client, author, pair, _phantom: Default::default() }
+	}
+}
```

Comment:
Currently, this design does not make that much sense because we need to instantiate the proposer every time, we call in the enclave to execute a slot. In the future, we will run the loop inside the enclave. So the factory does only need to be instantiated once in the beginning.

Reply:
We could wrap it in an `AtomicContainer` like I'm using for other components, to create the factory once and then access it from the global container.
But if we'll soon be moving the loop into the enclave, it might not be worth the effort.

Reply:
Yes, I think we should do that rather soon. As soon as we split parentchain sync and the execution of top in my opinion